File: typescript/session2/td/heritageMultiple/SuccRecursif.ts
import { EtatSucc } from "./EtatSucc";
import { AlgebreNatRecursif, AlgebreNatRecursif_defaut } from "./AlgebreNatRecursif";
import { Nat, FabriqueNat } from "../../../session1/td/naturels";
import { ZeroRecursif } from "./ZeroRecursif";
import { heriterInterfacesConcretes } from "../../../heritageMultiple/heritageInterfacesconcretes";

export class SuccRecursif extends EtatSucc implements AlgebreNatRecursif {
    public static readonly FAB: FabriqueNat<Nat> = new SuccRecursif(new ZeroRecursif());
    
    constructor(_predecesseur : Nat) {super(_predecesseur)}

    creerZero() {
        return new ZeroRecursif;
    }

    creerSuccesseur(x : Nat) : Nat {
        return new SuccRecursif(x);
    }

    estEgal(x : Nat) {
        return this.val() === x.val();
    }

    representation() : string {
        return String(this.val());
    }

    somme: (x: Nat) => Nat;
    produit: (x: Nat) => Nat;
	div: (x: Nat) => Nat;
	modulo: (x: Nat) => Nat;
	zero: () => Nat;
    un: () => Nat;
    val: () => number;
    estNul: () => boolean;
    predecesseur: () => Nat;
    taille: () => number;
    chiffre: (i: number) => number;
    creerNatAvecValeur: (valeur: number) => Nat;
    creerNatAvecRepresentation: (repDecimal: string) => Nat;
}
heriterInterfacesConcretes(SuccRecursif, [EtatSucc, AlgebreNatRecursif_defaut])

File: java/cours/src/session2/td/heritageMultiple/SuccRecursif.java
package session2.td.heritageMultiple;

import session1.td.Nat;

public class SuccRecursif extends EtatSucc implements AlgebreNatRecursifSuccesseur {
	public SuccRecursif(Nat predecesseur) {
		super(predecesseur);
	}

	@Override
	public Nat creerZero() {
		return new ZeroRecursif();
	}

	@Override
	public Nat creerSuccesseur(Nat predecesseur) {
		return new SuccRecursif(predecesseur);
	}

	@Override
	public String toString() {
		return this.representer();
	}

	@Override
	public boolean equals(Object obj) {
		return this.estEgal(obj);
	}
}
